analyses/ssd/flood_trigger/ssd_floodscan.md

```diff
@@ -0,0 +1,372 @@
+    return df_merged
+
+# Merge overlapping flood events
+# Each row in the input df should be an event
+# With start and end date columns: ['start_date'] and ['end_date']
+def merge_events(df):
+    df['flood_id'] = 0
+    f_id = 1
+    
+    # Loop through all of the events and tag the ones that are part of an overlap
+    for i in range(1, len(df.index)):        
+        start = df['start_date'].iloc[i,]
+        end = df['end_date'].iloc[i-1,]
+        if start < end:
+            df.loc[i, 'flood_id'] = f_id
+            df.loc[i-1, 'flood_id'] = f_id
+        else:           
+            df.loc[i-1, 'flood_id'] = f_id
+            f_id += 1
+    
+    # Now for each event, extract the min start data and max end date
+    df_start = df.groupby('flood_id')['start_date'].min().to_frame().reset_index()
+    df_end = df.groupby('flood_id')['end_date'].max().to_frame().reset_index()
+    
+    df_events = df_start.merge(df_end, on='flood_id').sort_values(by='start_date')
+    return df_events
+```
+
+```python
+def compute_flood_events(df_rainy,var_col,adm_list,outlier_thresh=3,adm_col='ADM2_EN',save_data=False):
+    list_floods_all=[]
+    for adm in adm_list:
+        df_floodscan_sel = df_rainy[df_rainy[adm_col]==adm].copy()
+        df_floodscan_sel['mean_cell_rolling'] = df_floodscan_sel[var_col].transform(lambda x: x.rolling(5, 1).mean())
+        df_floods_summary = df_floodscan_sel[(np.abs(stats.zscore(df_floodscan_sel['mean_cell_rolling'])) >= outlier_thresh)]
+        df_floods_summary = get_groups_consec_dates(df_floods_summary)
+        df_floods_summary = get_flood_summary(df_floods_summary)
+
+        df_summary_clean = merge_events(df_floods_summary)
+
+        if save_data: 
+            df_summary_clean.to_csv(country_data_exploration_dir / f'{adm}_floodscan_event_summary.csv', index=False)   
+        df_summary_clean[adm_col]=adm
+        list_floods_all.append(df_summary_clean)
+        df_floods_all=pd.concat(list_floods_all)
+        df_floods_all['year']=df_floods_all.start_date.dt.year
+    return df_floods_all
+```
+
+```python
+df_floods_all_median=compute_flood_events(df_floodscan_rainy,"median_ADM2_PCODE",adm2_list)
+```
+
+```python
+df_floods_all_mean=compute_flood_events(df_floodscan_rainy,"mean_ADM2_PCODE",adm2_list)
+```
+
+From the table and plot below we can see very different numbers of "flood events" per county. Since we define a flood event based on standard deviation, this means that in some counties there is a higher fluctuation than in others. I am not 100% sure whether this is a fair method to compare the different counties. 
+
+We also see that the different counties see "flood events" during different years. Where 2020, 2014, and 1998 saw the most counties with a flood event
+
+```python
+df_year=df_floods_all_median[['ADM2_EN','year']].drop_duplicates()
+```
+
+```python
+df_year.groupby("ADM2_EN").count()
+```
+
+```python
+df_year.groupby("year").count().sort_values('ADM2_EN',ascending=False)
```

Comment:
This seems like a large number of events. Are we aiming for a particular return period? Nationally or per county? Since 3 standard devations is 68% (assuming normality...) then I think that would correspond to 1/(1-0.68)= 1 in 3 year RP (per county). 

I guess I'm starting to worry that chopping things up by county is going to create some weird artefacts in the analysis. There are definitely counties that only have a few pixels.

Reply:
See conversation below but yes, 3SD was meant as 1 in 3 year return period, but didn't want to force a 1/3 RP since we don't know if flooding occurs in all counties. 
However, since we compute SD on all data within the rainy season, I don't think it is the best method either cause it means it will indicate floods more often in counties that have a lot of spikeness.. Do you have thoughts on what to best measure? 

>I guess I'm starting to worry that chopping things up by county is going to create some weird artefacts in the analysis.

Agree

Reply:
Long time ago, but yes agree. Looking at RP per year makes more sense